src/coq/serapi/processors/SerapiContentProcessor.js

```diff
@@ -0,0 +1,246 @@
+import SerapiProcessor from '../util/SerapiProcessor';
+import {
+  getGoalsFromResponse, getLastValidFullStop,
+  isGeneralMessage, parseErrorResponse, parseToSentence,
+} from '../SerapiParser';
+import * as Constants from '../SerapiConstants';
+import {byteIndexToStringIndex} from '../StringIndices';
+import {Mutex} from 'async-mutex';
+import {
+  createAddCommand, createCancelCommand, createGoalCommand,
+} from '../util/SerapiCommandFactory';
+
+class SerapiContentProcessor extends SerapiProcessor {
+  /**
+   *
+   * @param {SerapiTagger} tagger
+   * @param {SerapiState} state
+   * @param {EditorInterface} editor
+   */
+  constructor(tagger, state, editor) {
+    super(tagger, state, editor);
+    this.currentContent = '';
+    this.addLock = new Mutex();
+
+    this.contentLock = new Mutex();
+
+
+    this.addContent = '';
+  }
+
+  async setContent(content) {
+    console.log(`setContent(${content})`);
+    const releaseNewContent = await this.addLock.acquire();
+
+    this.addContent = content;
+
+    if (this.contentLock.isLocked()) {
+      releaseNewContent();
+      // TODO: what is the best way to return promises?
+      return Promise.resolve();
+    }
+
+
+    const releaseContentLock = await this.contentLock.acquire();
+    const stateRelease = await this.state.stateLock.acquire();
```

Comment:
Quite some code in these processors is devoted to the management of the state locks. This will make adding new processors more difficult. Would it be possible to move the state handling to a common part?